Create a JSON-RPC for KeY

wadoon opened the pull request:
# A remote KeY api as promised in KeY++

## Design Descisions

* We use handles to refer to large entities like, `KeYEnvironment`, `Proof`, or `Node`. 
  These handles are called `*Id` and are aligned hierarchical:
      
      EnvironemntId < ProofId < NodeId 

  If you have a `NodeId`, you can get a `ProofId` by calling `nodeId.proofId()`.


* We do not use expose any `key.core` classes. For example, Key's `Example`s are converted into `ExampleDesc` for the serialization. Every information holding class ends with `Desc`. 

* Given complex arguments (especially optional/nullable parameters) are packed into a class which ends with `Params`.

## TODO 

* [x] Implement a client in non-java, non-jvm (Python)

wadoon committed:
some functions for tff

# Conflicts:
#	settings.gradle

wadoon committed:
different clients

wadoon committed:
wip

wadoon committed:
Creating an JSON-RPC API

# Conflicts:
#	keyext.api/src/main/java/org/keyproject/key/api/FileTypeAdapter.java
#	keyext.api/src/main/java/org/keyproject/key/api/remoteapi/EnvId.java
#	keyext.api/src/main/java/org/keyproject/key/api/remoteapi/KeyApiImpl.java
#	keyext.api/src/main/java/org/keyproject/key/api/remoteapi/LoadParams.java
#	keyext.api/src/main/java/org/keyproject/key/api/remoteapi/ProofId.java
#	keyext.api/src/main/java/org/keyproject/key/api/remoteapi/ProofLoading.java

# Conflicts:
#	key.core/src/main/java/de/uka/ilkd/key/proof/Proof.java

# Conflicts:
#	key.core/src/main/java/de/uka/ilkd/key/proof/Proof.java

wadoon committed:
working on a working minimal version

wadoon committed:
more doc and py generation

Drodt commented on api.meta.md:
````diff
@@ -0,0 +1,480 @@
+```
+
+
+### client/sm (`server ~~> client`) 
````
I would prefer verbose names. Avoid abbreviations like "sm".

wadoon replied:
Yes, something from the beginning. 

I am still very unhappy with the client API. It is mainly the progress listener, which is not very good, especially with parallel processes.

Drodt commented on api.meta.md:
````diff
@@ -0,0 +1,480 @@
+### client/sm (`server ~~> client`) 
+
+```
+Client.client/sm( arg0 : ShowMessageParams ) **async**
````
Similarly, avoid generic argument names like `arg0`.

wadoon replied:
Sorry, this is due to the lack of Java reflection. I think I need to compile with `-g` for debugging information. 

These documentation are generated from the Java information.

wadoon replied:
Since Java 8, the `-parameters` switch ensures the parameter names in class files.

Drodt commented on key.core.symbolic_execution/src/test/java/de/uka/ilkd/key/symbolic_execution/testcase/AbstractSymbolicExecutionTestCase.java:
```diff
@@ -906,15 +904,15 @@ protected static void assertConstraint(IExecutionConstraint expected,
-     * @param expected The expected node.
-     * @param current The current node.
-     * @param compareVariables Compare variables?
+     * @param expected           The expected node.
```
Is this a new spotless setting?

Drodt reviewed (commented):
I like the approach. It seems like a good setup.

For what languages do you want to write a client? What's on your TODO list?

wadoon committed:
running for primitive data types, somethings wrong in de-/serialization

wadoon commented:
> For what languages do you want to write a client? What's on your TODO list?

I make a simple one in Python. The Python stubs are mostly generated by using Java Reflection. I am creating a JSON with meta information, markdown and Python classes. 

I don't know whether you want a rust client. In general, you need a background thread for reading messages, and a shared map from request ids to wait conditions/locks for block synchronous requests until return value is received. 

The Python API is more for testing, I am not planning a production-worthy state. 

Do we have a use case? For JVM programs I would still recommend to use KeY directly.

wadoon committed:
add throwable adapter

wadoon committed:
spotless and merge errors

wadoon committed:
fix after refactoring + spotless

wadoon committed:
wip

wadoon committed:
start of a simple Java client

samysweb committed:
- Shutdown Callbacks
- TreeNodeDesc
- Some stabilization of features

# Conflicts:
#	keyext.api/src/main/java/org/keyproject/key/api/StartServer.java

samysweb committed:
Error message for python generation

samysweb committed:
Updated library and example

samysweb committed:
Added proof api endpoint

samysweb committed:
Attempt at reseting proof control

samysweb committed:
First (incomplete) working proof management endpoint

samysweb committed:
Updated python client for proof endpoint

samysweb committed:
Improved interface, environment disposal

samysweb committed:
Fixed JSON type annotation&retrieval

samysweb committed:
Updated json type annotation on codegen/python side

samysweb committed:
Merge branch 'steuber/jsonrpc' into weigl/jsonrpc

samysweb committed:
Configurable Strategy (not Streategy) options for proof auto mode

samysweb committed:
Extracted node description function

samysweb committed:
Null serialization (necessary due to python class initialization), node labels

wadoon committed:
reformat / remove weakrefs from the data model

wadoon commented:
*KaKeY:* Make this state fit for merging.

wadoon committed:
fix compile error

wadoon committed:
spotless

wadoon committed:
fixing the missing javadoc in the api generator

wadoon committed:
better documentation

wadoon committed:
working on documentation, now with json examples

# Conflicts:
#	key.ui/src/main/java/de/uka/ilkd/key/gui/ExampleChooser.java

wadoon committed:
try to make the documentation better

wadoon committed:
better documentation

wadoon committed:
repair branch after rewriting

unp1 committed:
Extensions to the KeY API: Server changes allow now a reconnect

unp1 commented:
I pushed some minot changes to fix some compilation bugs and enhanced the server side to allow for reconnects. The Java API work now (besides the rusty and solidity KeY versions) with KeY-tau (https://github.com/KeYProject/key-tau)

unp1 committed:
Fix server API name and example chooser

wadoon commented:
@unp1 Please, no merge of main into PRs. This makes things really complicated.

wadoon committed:
spotless

wadoon commented:
KaKeY: 2026-06-19 -- Externalize

wadoon commented:
Externalized into https://github.com/KeYProject/key-rpc
Remaining parts in #3849 

/close
